doc/manual/methods.rst

```diff
@@ -552,4 +558,164 @@ arguments::
+
+Below we discuss particular challenges and some alternative ways to resolve such issues.
+
+Tuple and NTuple arguments
+~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``Tuple`` (and ``NTuple``) arguments present special challenges. For example,
+::
+
+   f{N}(x::NTuple{N,Int}) = 1
+   f{N}(x::NTuple{N,Float64}) = 2
+
+are ambiguous because of the possibility that ``N == 0``. To resolve the
+ambiguity, one approach is define a method for the empty tuple::
+
+  f(x::Tuple{}) = 3
+
+Alternatively, you can insist that there is at least one element in the tuple::
+
+  f(x::Tuple{Float64, Vararg{Float64}}) = 2
+
+and let the ``NTuple{N,Int}`` method be the fallback.
+
+Use generic single-argument wrapper functions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+When you might be tempted to dispatch on two or more arguments,
+consider whether a "wrapper" function might make for a simpler
+design. For example, instead of writing multiple variants::
+
+  f(x::A, y::A) = ...
+  f(x::A, y::B) = ...
+  f(x::B, y::A) = ...
+  f(x::B, y::B) = ...
+
+you might consider defining
+::
+
+   f(x::A, y::A) = ...
+   f(x, y) = f(g(x), g(y))
+
+where ``g`` converts the argument to type ``A``. Most likely, ``g`` will
+need a fallback definition
+::
+
+   g(x::A) = x
+
+
+Dispatch on one argument at a time
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+If you need to dispatch on multiple arguments, and there are many
+fallbacks with too many combinations to make it practical to define
+all possible variants, then consider introducing a "name cascade"
+where (for example) you dispatch on the first argument and then call
+an internal method::
+
+  f(x::A, y) = _fA(x, y)
+  f(x::B, y) = _fB(x, y)
+
+Then the internal methods ``_fA`` and ``_fB`` can dispatch on ``y`` without
+concern about ambiguities with each other with respect to ``x``.
+
+Be aware that this strategy has at least one major disadvantage: in
+many cases, it is not possible for users to further customize the
+behavior of ``f`` by defining further specializations of your exported
+function ``f``. Instead, they have to define specializations for your
+internal methods ``_fA`` and ``_fB``, and this blurs the lines between
+exported and internal methods.
+
+Abstract containers and element types
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Where possible, try to avoid defining methods that dispatch on element types of abstract
+containers. For example,
+::
+
+   -{T<:Date}(A::AbstractArray{T}, b::Date)
+
+will generate ambiguities for anyone who defines a method
+::
+
+   -{T}(A::MyArrayType{T}, b::T)
+
+In order of preference, some strategies are:
+
+- avoid defining either of these methods: instead, rely on a generic method ``-{T}(A::AbstractArray{T}, b)`` and make sure this method is implemented with generic calls (like ``similar`` and ``-``) that do the right thing for each container type and element type *separately*.
+
+- if at least one of these must be defined, prioritize dispatch on the container
```

Comment:
That is to say, where possible don't dispatch on `T`, but it's OK (often necessary) to dispatch on `MyArray`.